src/libraries/System.Private.CoreLib/src/System/IO/File.cs

```diff
@@ -901,5 +926,42 @@ private static long GetPreallocationSize(FileMode mode, string? contents, Encodi
 
             return preambleSize + encoding.GetByteCount(contents);
         }
+
+        private static async IAsyncEnumerable<string> IterateFileLinesAsync(DisposingFlagStreamReader sr, string path, Encoding encoding, CancellationToken ctEnumerable, [EnumeratorCancellation] CancellationToken ctEnumerator = default)
+        {
+            if (sr.IsDisposed)
```

Comment:
in what scenario is it possible for the `DisposingFlagStreamReader ` to be disposed here?

Reply:
> PTAL at my comments, I would like to get a better understanding of why the `StreamReader` can be disposed an re-created more than once

Answer together here :)
> in what scenario is it possible for the `DisposingFlagStreamReader ` to be disposed here?

Before start this code, I checked existing IEnumerable<string> File.ReadLines() [related](https://github.com/dotnet/runtime/blob/4ed596ef63e60ce54cfb41d55928f0fe45f65cf3/src/libraries/System.Private.CoreLib/src/System/IO/ReadLinesIterator.cs#L11) code and related [comment](https://github.com/dotnet/runtime/pull/43108#discussion_r510916526) and got some background that we would move first `streamReader` allocation before involve `Enumerator` (so pass streamReader to this `IterateFileLinesAsync()`) so that exceptions such as DirectoryNotFoundException and FileNotFoundException are thrown directly by `File.ReadLinesAsync` (which the user probably expects), rather than throw until call `Enumerator.MoveNext`.
And there was [comment](https://github.com/dotnet/runtime/pull/66492#discussion_r825538143) to suggest using compiler-generated Iterator to implement. 

Because generated enumerable stores first time passed streamReader **reference**, so here we check if current streamReader reference has been disposed already for scenario: when previous user has consumed all lines and then call `GetEnumerator` against same enumerable).

```csharp
            IAsyncEnumerable<string> enumerable = File.ReadLinesAsync("path");
            await foreach (string line in enumerable)
            {
                // consume line;
            } // streamReader is disposed now

            // can use same enumerable to create another enumerator and work properly
            await foreach (string line in enumerable)
            {
                // consume line;
            }
```

Reply:
@lateapexearlyspeed thanks for the explanation! Now I can see that we do something similar for `File.ReadLines`:

https://github.com/dotnet/runtime/blob/a077f271c3b5a8487f94111dc00a404a01fdc86b/src/libraries/System.Private.CoreLib/src/System/IO/ReadLinesIterator.cs#L64-L70

I wonder whether we need a new type for `DisposingFlagStreamReader`. The `ReadLinesAsync` method could create a `FileStream` and pass it here and then we could use `FileStream.CanRead` to figure out whether it has been disposed or not:

https://github.com/dotnet/runtime/blob/a077f271c3b5a8487f94111dc00a404a01fdc86b/src/libraries/System.Private.CoreLib/src/System/IO/Strategies/OSFileStreamStrategy.cs#L73

Reply:
@stephentoub just to be extra sure I assume that we want to have this behavior?

Reply:
Fixed by using `FileStream.CanRead`. Use `StreamReader.BaseStream` to get inner fileStream so can pass and reuse StreamReader method to read lines.